app/src/main/java/org/onionshare/android/ShareManager.kt

```diff
@@ -1,221 +1,155 @@
-        if (newList.isEmpty()) {
-            _shareState.value = ShareUiState.NoFiles
+        // initial state: Adding file and services stopped
+        if (f is FilesState.Added && t is TorState.Stopped && w is WebServerState.Stopped) {
+            if (f.files.isEmpty()) emit(ShareUiState.NoFiles)
+            else emit(ShareUiState.FilesAdded(f.files))
+        } // handle error while adding files
+        else if (f is FilesState.Error) {
+            emit(ShareUiState.ErrorAddingFile(f.files, f.errorFile))
+            // special case handling for error state without file left
+            if (f.files.isEmpty()) {
+                delay(1000)
+                emit(ShareUiState.NoFiles)
+            }
+        } // continue with zipping and report state while doing it
+        else if (f is FilesState.Zipping && t is TorState.Starting) {
+            val torPercent = (t as? TorState.Starting)?.progress ?: 0
+            emit(ShareUiState.Starting(f.files, f.progress, torPercent))
+        } // after zipping is complete, and webserver still stopped, start it
+        else if (f is FilesState.Zipped && !shouldStop &&
+            (t is TorState.Starting || t is TorState.Started) && w is WebServerState.Stopped
+        ) {
+            webserverManager.start(f.sendPage)
+            val torPercent = (t as? TorState.Starting)?.progress ?: 0
+            emit(ShareUiState.Starting(f.files, 100, torPercent))
+        } // continue to report Tor progress after files are zipped
+        else if (f is FilesState.Zipped && t is TorState.Starting) {
+            emit(ShareUiState.Starting(f.files, 100, t.progress))
+        } // everything is done, show sharing state with onion address
+        else if (f is FilesState.Zipped && t is TorState.Started && w is WebServerState.Started) {
+            val url = "http://${t.onion}.onion"
+            emit(ShareUiState.Sharing(f.files, url))
+            notificationManager.onSharing()
+        } // if webserver says download is complete, report that back
+        else if (w is WebServerState.DownloadComplete) {
+            stopSharing()
+            emit(ShareUiState.Complete(f.files))
+        } // handle stopping state
+        else if (t is TorState.Stopping) {
+            emit(ShareUiState.Stopping(f.files))
+        } // handle unexpected stopping/stopped only after zipped, because we start webserver only when that happens
+        else if (!shouldStop && f is FilesState.Zipped &&
+            (t is TorState.Stopping || t is TorState.Stopped || w is WebServerState.Stopped)
+        ) {
+            emit(ShareUiState.Error(f.files))
         } else {
-            _shareState.value = ShareUiState.FilesAdded(newList)
-        }
-    }
-
-    fun removeAll() {
-        // release persistable Uri permissions again
-        shareState.value.files.forEach { file ->
-            file.releaseUriPermission()
+            LOG.error("Unhandled state: ↑")
```

Comment:
This entire `shareState` method is quite a beast and does not handle all permutations of states. If there's any suggestions how to reconcile the states of the various sub-systems in a better way, please let me know!